CANTroller2/src/devices.h

```diff
@@ -596,171 +591,177 @@ class Speedometer : public PulseSensor<float> {
-//         // attachInterrupt(digitalPinToInterrupt(pin), isr, RISING);
-//     }
-//     void isr(void) {  //  A better approach would be to read, reset, and restart a hardware interval timer in the isr.  Better still would be to regularly read a hardware binary counter clocked by the pin - no isr.
-//         int32_t temp_us = PulseTimer.elapsed();
-//         if (temp_us > delta_impossible_us) {
-//             delta_us = temp_us;    
-//             PulseTimer.reset();
-//         }
-//     }
-//     void calc() {
-//         if (val_source != _TOUCH && val_source != _POT) {
-//             float val_temp;
-//             if (PulseTimer.elapsed() < stop_timeout_us) {
-//                 if (delta_us <= 0) printf ("Warning: PulseSensor::calc sees delta_us <= 0\n");
-//                 else val_temp = conversion_factor/delta_us;
-//             }
-//             else val_temp = 0;     
-//             if (filt_lp_spike(val_temp)) {
-//                 if (val_temp == 0) assign_val(0.0);
-//                 else assign_val ( filt_ema(val_temp, *d_val, ema_alpha) );
-//             }
-//         }
-//     }
-// };
-// // Device::Transducer::Sensor::InPWM are servo-pwm standard signals being read in, with variable pulsewidth
-// class InPWM : public Sensor {
-//   protected:
-//     Timer PulseTimer;  // OK to not be volatile?
-//   public:
-//     InPWM(int32_t arg_pin, bool arg_dir, float arg_val_min, float arg_val_max, float arg_val_cent)
-//     : Sensor(arg_pin, arg_dir, arg_val_min, arg_val_max, arg_val_cent)  {}
-
-// };
-// // Device::Transducer::Sensor::InPWM::InPWMToggle are servo-pwm standard signals being read in, but only valid values are pulse_min (0) and pulse_max (1) (eg hotrc ch3-4)
-// class InPWMToggle : public InPWM {
-//   protected:
-//     Timer PulseTimer;  // OK to not be volatile?
-//   public:
-//     InPWMToggle(int32_t arg_pin, bool arg_dir, float arg_val_min, float arg_val_max, float arg_val_cent)
-//     : Sensor(arg_pin, arg_dir, arg_val_min, arg_val_max, arg_val_cent)  {}
-
-// };
-
-// NOTE: I implemented the gas servo, but it looks like it's all in native units. should it still be a transducer?
 // ServoPWM is a base class for our type of actuators, where by varying a pulse width (in us), motors move.
 //    e.g. the gas, brake and steering motors. The gas motor is an actual servo, the others are controlled with servo signaling via jaguars.
-template<typename NATIVE_T, typename HUMAN_T>
-class ServoPWM : public Transducer<NATIVE_T, HUMAN_T> {
+template<typename HUMAN_T>
+class ServoPWM : public Transducer<int32_t, HUMAN_T> {
   protected:
     Servo _servo;
-
-    // NOTE: should be marked 'override' but compiler says it doesn't override anything...?
-    void set_native_limits(Param<NATIVE_T> &minParam, Param<NATIVE_T> &maxParam) {
-        this->set_native_limits(minParam, maxParam);
-        _servo.attach(this->_pin, this->min_native->get(), this->max_native->get());
+    virtual void handle_pin_mode() {
```

Comment:
write()? Seems like we're using this to write to the servo, based on this line and line 605

Reply:
we still need this, since it's how we respond to being in pin mode when update() is called, but i can add a write() function that does the actual writing (that way we could call it directly if we ever needed to).